pkg/planner/core/indexmerge_path.go

```diff
@@ -846,6 +1254,66 @@ func (ds *DataSource) collectFilters4MVIndex(filters []expression.Expression, id
 	return accessFilters, remainingFilters
 }
 
+// For idx(x, cast(a as array), z), `x=1 and (2 member of a) and (1 member of a) and z=1 and x+z>0` is split to:
+// accessFilters combination:
+// 1: `x=1 and (2 member of a) and z=1`, remaining: `x+z>0`.
+// 2: `x=1 and (1 member of a) and z=1`, remaining: `x+z>0`.
+//
+// just as the 3rd point as we said in generateIndexMerge4ComposedIndex
+//
+// 3: The predicate of mv index can not converge to a linear interval range at physical phase like EQ and
+// GT in normal index. Among the predicates in mv index (member-of/contains/overlap), multi conditions
+// about them should be built as self-independent index path, deriving the final intersection/union handles,
+// which means a mv index path may be reused for multi related conditions. Here means whether (2 member of a)
+// And (1 member of a) is valid composed range or empty range can't be told until runtime intersection/union.
+//
+// therefore, for multi condition about a single mv index virtual json col here: (2 member of a) and (1 member of a)
+// we should build indexMerge above them, and each of them can access to the same mv index. That's why
+// we should derive the mutations of virtual json col's access condition, output the accessFilter combination
+// for each mutation of it.
+
+// todo: case like idx(x, cast(a as array), z), condition like: x=1 and x=2 and (
+// 2 member of a)? we can derive the x is invalid range?
+func (ds *DataSource) collectFilters4MVIndexMutations(filters []expression.Expression,
+	idxCols []*expression.Column) (accessFilters, remainingFilters []expression.Expression, mvColOffset int, mvFilterMutations []expression.Expression) {
+	usedAsAccess := make([]bool, len(filters))
+	// accessFilters [x, a<json>, z]
+	//                    |
+	//                    +----> it may have several substitutions in mvFilterMutations if it's json col.
+	mvFilterMutations = make([]expression.Expression, 0, 1)
+	mvColOffset = -1
+	for z, col := range idxCols {
+		found := false
+		for i, f := range filters {
+			if usedAsAccess[i] {
+				continue
+			}
+			if ds.checkFilter4MVIndexColumn(f, col) {
+				// todo: ban the virtual col in mv index except the json col it self.
+				if col.VirtualExpr != nil {
+					// assert jsonColOffset should always be the same.
+					// if the filter is from virtual expression, it means it is about the mv json col.
+					mvFilterMutations = append(mvFilterMutations, f)
+					mvColOffset = z
```

Comment:
This line contains an implicit assumption: there should be only 1 virtual column in a MVIndex? Can we add this into the comment?

Reply:
I changed the if condition to: `if col.VirtualExpr != nil && col.VirtualExpr.GetType().IsArray() `